sys/net/application_layer/gcoap/forward_proxy.c

```diff
@@ -277,29 +337,27 @@ static int _gcoap_forward_proxy_copy_options(coap_pkt_t *pkt,
         if (optlen >= 0) {
             if (IS_USED(MODULE_NANOCOAP_CACHE) && !etag_added && (opt.opt_num >= COAP_OPT_ETAG)) {
                 static const uint8_t tmp[COAP_ETAG_LENGTH_MAX] = { 0 };
-                /* add slack to maybe add an ETag on stale cache hit later, as is done in gcoap_req_send()
-                 * (which we circumvented in _gcoap_forward_proxy_via_coap()) */
+                /* add slack to maybe add an ETag on stale cache hit later, as is done in
+                 * gcoap_req_send() (which we circumvented in _gcoap_forward_proxy_via_coap()) */
                 if (coap_opt_add_opaque(pkt, COAP_OPT_ETAG, tmp, sizeof(tmp))) {
                     etag_added = true;
                 }
             }
-#if IS_USED(MODULE_NANOCOAP_CACHE)
-            /* req_tag in cep is pre-processor guarded so we need to as well */
-            if (opt.opt_num == COAP_OPT_ETAG) {
-                if (cep->req_etag_len == 0) {
+            if (IS_USED(MODULE_NANOCOAP_CACHE) && opt.opt_num == COAP_OPT_ETAG) {
+                if (_cep_get_req_etag_len(cep) == 0) {
                     /* TODO: what to do on multiple ETags? */
-                    cep->req_etag_len = (uint8_t)optlen;
+                    _cep_set_req_etag_len(cep, (uint8_t)optlen);
+#if IS_USED(MODULE_NANOCOAP_CACHE)
+                    /* req_tag in cep is pre-processor guarded so we need to as well */
                     memcpy(cep->req_etag, value, optlen);
+#endif
```

Comment:
this doesn't make sense to me. Why is the memcpy extra guarded when it already can not be reached due to the check in L346?

Reply:
Because `req_etag` is only available, when the `nanocoap_cache` module is used. The check in 346 is effectively done at runtime (even though the compiler will optimize out the static `true`/`false`, while this guard is checked at compile time. If the guard would not be there, the compiler would fail if `nanocoap_cache` is not present. Sure, we can go back to the guard around the whole `if` as it was before this change, but then other stuff is only checked by the compile if `nanocoap_cache` is present (and personally I find the non-CPP use of `IS_USED` more readable and try to avoid the CPP-use if possible).